lib/redis/pool.js

```diff
@@ -0,0 +1,96 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+const genericPool = require('generic-pool')
+const P = require('../promise')
+const redis = require('redis')
+const redisConnection = require('./connection')
+
+P.promisifyAll(redis.RedisClient.prototype)
+P.promisifyAll(redis.Multi.prototype)
+
+module.exports = (config, log) => {
+  const redisConfig = {
+    host: config.host,
+    port: config.port,
+    prefix: config.sessionsKeyPrefix,
+    // Prefer redis to fail fast than wait indefinitely for reconnection
+    enable_offline_queue: false
+  }
+
+  const redisFactory = {
+    create () {
+      return new P((resolve, reject) => {
+        let connection
+
+        const client = redis.createClient(Object.assign({}, redisConfig, {
+          retry_strategy ({ attempt }) {
+            if (attempt <= config.retryCount) {
+              return config.initialBackoff * Math.pow(2, attempt - 1)
+            }
+
+            if (connection) {
+              // It's too late to reject here because the connection was
+              // already added to the pool. Destroy it instead.
+              connection.destroy()
+            } else {
+              reject(new Error('Redis connection failed'))
+            }
+          }
+        }))
+
+        client.on('ready', () => {
+          if (! connection) {
+            connection = redisConnection(log, client)
+            resolve(connection)
+          }
+        })
+
+        client.on('error', err => {
+          log.error({ op: 'redis.error', err: err.message })
```

Comment:
I don't think so, although there's probably no harm in doing it if you think we should. In practice this error isn't really that useful to us because it appears to fire in two situations:

1. If we reached the failure condition in `retry_strategy`, at which point the connection has already been destroyed. Note that before I implemented `retry_strategy`, this error was not firing for connection issues *at all*.

2. If there's any kind of logical error inside the library, e.g. invalid arguments to one of the methods. In these cases we don't actually need to mark the connection itself as invalid and, anyway, I don't think there's any paths in our codebase that trigger them (although I may be mistaken).

I'm mostly logging here for visibility just in case it ever starts happening.